docs/Disaster Recovery Plan.md

```diff
@@ -0,0 +1,319 @@
+
+### Record Latest Known Good State
+
+> 1. [Establish](#record-state-task-1) what the last known good state is on chain and match it up to
+>    the sequencer's database.
+> 2. [Record](#record-state-task-2) this state (the last inserted identity and the last valid root)
+>    carefully for reference.
+
+The kinds of issues that can trigger use of this procedure may involve data corruption on chain.
+This means that in order to restore service properly it needs to be very clear what the last known
+good state is.
+
+> #### Record State: Task 1
+>
+> 1. Obtain read access to the production database via the VPN already obtained.
+> 2. Find the row in the signup sequencer database that corresponds to the last correct root on
+>    chain (hereafter `rootValue`). Note that this may not be the same as `latestRoot` as data may
+>    have become corrupt.
+>
+> ```sql
+> SELECT *                -- Select all columns
+> FROM root_history       -- From the root history table
+> WHERE root = rootValue; -- In rows that match the root value
+> ```
+>
+> 3. Find the last identity inserted as part of the root returned from the above.
+>
+> ```sql
+> SELECT *                          -- Select all columns
+> FROM identities                   -- From the identities table
+> WHERE commitment = last_identity; -- In rows where commitment matches the last_identity from the
+>                                   -- previous query
+> ```
+
+This gives us the last known good state that the contract was in, and hence the point from which any
+data restoration effort needs to begin.
+
+> #### Record State: Task 2
+>
+> Record this state as:
+>
+> - **A row in the identities table** of the signup sequencer database; in particular, this is the
+>   last known identity where `status == "mined"` that is known to have been correctly submitted on
+>   chain.
+> - **A row in the roots table** of the signup sequencer database; in particular, this is the last
+>   known root where `status == "mined"` that is known to have been correctly submitted on chain.
+
+Once this is done you can be assured that all of the necessary state is available for restoring the
+operation of the contract on chain. Make sure to keep this data handy, as it will be necessary for
+multiple future steps.
+
+### Roll-Back the Signup Sequencer
+
+> 1. Obtain write access to the production database from the infrastructure team.
+> 2. [Back up](#roll-back-task-2) the identities and root history tables.
+> 3. [Roll back](#roll-back-task-3) the signup sequencer database such that any identity or root
+>    with `status == "mined"` that is not properly represented on the chain is back in
+>    `status == "pending"`.
+
+Unfortunately it is _possible_ that the signup sequencer record the status of identities and roots
+as `"mined"` when this is spurious (due to on-chain bugs or corruption). To that end we have to be
+able to roll back the database to a state that correctly matches the on-chain state to be restored.
```

Comment:
Is there any way to find the latest root that's common to both the signup-sequencer and the contract?

Reply:
Unfortunately not. In the presence of data corruption we can basically only rely on the signup sequencer.